capsules/src/process_console.rs

```diff
@@ -1147,28 +1185,39 @@ impl<'a, C: ProcessManagementCapability> uart::ReceiveClient for ProcessConsole<
             match rx_len {
                 0 => debug!("ProcessConsole had read of 0 bytes"),
                 1 => {
-                    self.command_buffer.map(|command| {
-                        let index = self.command_index.get() as usize;
-                        if read_buf[0] == ('\n' as u8) || read_buf[0] == ('\r' as u8) {
-                            self.execute.set(true);
-                            let _ = self.write_bytes(&['\r' as u8, '\n' as u8]);
-                        } else if read_buf[0] == ('\x08' as u8) && index > 0 {
-                            // Backspace, echo and remove last byte
-                            // Note echo is '\b \b' to erase
-                            let _ = self.write_bytes(&['\x08' as u8, ' ' as u8, '\x08' as u8]);
-                            command[index - 1] = '\0' as u8;
-                            self.command_index.set(index - 1);
-                        } else if index < (command.len() - 1) && read_buf[0] < 128 {
-                            // For some reason, sometimes reads return > 127 but no error,
-                            // which causes utf-8 decoding failure, so check byte is < 128. -pal
-
-                            // Echo the byte and store it
-                            let _ = self.write_byte(read_buf[0]);
-                            command[index] = read_buf[0];
-                            self.command_index.set(index + 1);
-                            command[index + 1] = 0;
+                    if !self.first_prompt_shown.get() {
+                        if self.print_welcome_message.get() {
+                            self.display_welcome();
+                        } else {
+                            self.prompt();
                         }
-                    });
+                    } else {
+                        self.command_buffer.map(|command| {
+                            let index = self.command_index.get() as usize;
+                            if read_buf[0] == ('\n' as u8) || read_buf[0] == ('\r' as u8) {
+                                self.execute.set(true);
+                                let _ = self.write_bytes(&['\r' as u8, '\n' as u8]);
+                            } else if read_buf[0] == ('\x08' as u8) {
+                                if index > 0 {
```

Comment:
why move this if down to the next line if you don't do anything else outside of it?

Reply:
I'm not sure I understand the question. Do you mean why index > 0 is not placed within the upper if?

Reply:
Yeah that is what I mean. I only ask because I see you changed it but as far as I can tell the behavior is unchanged, so I wasn't sure if there was a mistake

Reply:
Without the if, if a user presses backspace more than the number of characters he wrote, the console "deletes" whatever text was written before the user started writing the command. Most likely, he will detele the prompt text.

Reply:
Oh, I understand now, sorry for the noise